python/ray/experimental/channel/cpu_nccl_group.py

```diff
@@ -0,0 +1,273 @@
+        self.communicators.add(comm)
+
+        ray.get(comm.wait_p2p.remote(self.num_ops[comm_key], tensor))
+        self.num_ops[comm_key] += 1
+
+    def recv(
+        self,
+        shape: Tuple[int],
+        dtype: torch.dtype,
+        peer_rank: int,
+        allocator: Optional[TorchTensorAllocator] = None,
+    ):
+        """Receive the tensor from the communicator actor."""
+        comm_key = f"communicator-p2p-{peer_rank}-{self.get_self_rank()}"
+        comm = CPUCommunicator.options(name=comm_key, get_if_exists=True).remote(2)
+        self.communicators.add(comm)
+
+        received_tensor = ray.get(comm.wait_p2p.remote(self.num_ops[comm_key]))
+        assert (
+            allocator is not None
+        ), "torch tensor allocator is required for CPUNcclGroup"
+        buf = allocator(shape, dtype)
+        buf[:] = received_tensor[:]
+        self.num_ops[comm_key] += 1
+        return buf
+
+    def allreduce(
+        self,
+        send_buf: torch.Tensor,
+        recv_buf: torch.Tensor,
+        op: ReduceOp = ReduceOp.SUM,
+    ):
+        all_ranks = [
+            self.get_rank(actor_handle) for actor_handle in self.get_actor_handles()
+        ]
+        comm_key = "communicator-collective-" + "-".join(map(str, sorted(all_ranks)))
+        comm = CPUCommunicator.options(name=comm_key, get_if_exists=True).remote(
+            self._world_size
+        )
+        self.communicators.add(comm)
+
+        result = ray.get(
+            comm.wait_collective.remote(self.num_ops[comm_key], send_buf, op)
+        )
+        assert recv_buf is not None, "Receiving buffer required for CPUNcclGroup"
+        recv_buf[:] = result[:]
+        self.num_ops[comm_key] += 1
+
+    def destroy(self) -> None:
+        for communicator in self.communicators:
+            ray.kill(communicator)
+
+    def initialize(self, rank: int) -> None:
+        self._rank = rank
+
+    def get_actor_handles(self) -> List["ray.actor.ActorHandle"]:
+        return self._actor_handles
+
+    def get_rank(self, actor: ray.actor.ActorHandle) -> int:
+        """
+        Return the given actor's rank in the NCCL communicator.
+
+        Args:
+            actor: The actor handle to look up.
+        """
+        actor_ids = [a._ray_actor_id for a in self._actor_handles]
+        try:
+            rank = actor_ids.index(actor._ray_actor_id)
+        except ValueError:
+            raise ValueError("Actor is not in the NCCL group.")
+        return rank
+
+    def get_self_rank(self) -> Optional[int]:
+        return self._rank
+
+    def get_world_size(self) -> int:
+        """
+        Return the number of ranks in the NCCL communicator.
+        """
+        return self._world_size
+
+    def recv_stream(self):
+        raise NotImplementedError
+
+    def send_stream(self):
+        raise NotImplementedError
+
+
+class MockCudaStream:
+    def __init__(self):
+        self.cuda_stream = 0
+
+
+def start_nccl_mock():
```

Comment:
Is this only for testing purposes? It'd be good to move it to a test file or a utils file in `dag/tests/experimental`, to keep it separate from user-facing code.